### ISSUE TITLE
Feat: Client-side JS logging setup

### ISSUE THREAD

**Issue by yregaieg** [Alfresco/Aikau#530]
Hey there,
  I am currently using the "alfresco/services/LoggingService" to log client side activity.
As you must know, client-side logging can be a performance pit even if you setup log filters, so sometimes I need to deactivate logging on all pages. I can of course rely on maven profiles to setup my logging preference during build and then I will need to redeploy my webapp, but it would be really nice if I could configure : 
* Default logging preference to which the LoggingService should fallback in case of absence of specific configuration
* Whether or not specific config in pages should override the default logging preference

in a property file

**Comment by draperd** [comment]
I saw that this came up on the IRC chat... it is possible to configure the alfresco/logging/Service so that it has a set of default preferences. We do this in all our unit test application pages, e.g.

```JAVASCRIPT
{
   name: "alfresco/services/LoggingService",
   config: {
      loggingPreferences: {
         enabled: true,
         all: true,
         warn: true,
         error: true
      }
   }
}
```

However, it's worth bearing in mind that the LoggingService is only intended to be used for development purposes and shouldn't be used in production. In fact, when client-debug mode is disabled logging will not even occur, therefore we don't consider the performance impact particularly relevant.

If you want to capture client-side errors on a production system you could consider including the "alfresco/services/ErrorReporter" that will log client-side errors on the server.

My understanding from Ole of the IRC conversation was that a requirement was to prevent individual user preferences for logging from overriding the defaults (which we can do as a feature request).

I don't necessarily understand the benefits of setting a default logging preference... presumably if there is no configuration you just want everything enabled? 

Unfortunately in order to retain backwards compatibly from release to release we only ever provide new features via opt-in configuration. E.g. if someone is previously using the LoggingService without specific configuration enabled then this would change their previous behaviour when upgrading to a new Aikau release.

As for storing the information in a property file, that actually falls out of the scope of Aikau - referencing the property file would need to be done via the WebScript JavaScript controller and that is in the hands of the developer making use of Aikau - unless you can think of an alternative approach for accessing a properties files?

I will raise a feature though to address preventing user preferences from overriding LoggingService configuration though.

**Comment by draperd** [comment]
I've raised https://issues.alfresco.com/jira/browse/AKU-578 to implement config taking precedence over user preferences. It's only a small change so I'll try and include it in the current sprint if possible.

**Comment by yregaieg** [comment]
Hey Dave,
Thanks for the quick reply !

>As for storing the information in a property file, that actually falls out of the scope of Aikau - referencing the property file would need to be done via the WebScript JavaScript controller and that is in the hands of the developer making use of Aikau - unless you can think of an alternative approach for accessing a properties files?

 I spent the last hour trying to find out a lead on how is aikau retrieving information like username, context urls, .... in AlfConstants, but couldn't find any thing would you please provide a lead for that?
 Is it possible to implement the same logic to bring some params from a property file to some widget or service ?

I do not know why I've never considered it, until now, but I am thinking I can extend the logging service to something tightly adapted to my needs and then it will be possible for me to make system wide preferences by just altering my new service accordingly.

**Comment by draperd** [comment]
All of the constants are provided by a WebScript in Surf: code/root/projects/surf/spring-surf/spring-surf/src/main/resources/webscripts/amd-modules/UrlConstants.get.js.ftl

We've configures an AMD package called "service" which can call WebScripts that return JavaScript module (although we only have this one at the moment).

You can reference it in your own module by placing a dependency on "service/constants/Default". This returns a module that is a simple JavaScript object containing a whole bunch of application specific stuff (e.g. current user name, the debug mode state, URL prefixes, etc). 

This is because it's application context related code that isn't readily available on the client. You could quite easily create something similar to this if you wanted to access any configuration from your server.

I hope that helps.

Also... you're right that extending the LoggingService (or even just creating your own service from scratch) is a perfectly reasonable thing to do. As far as I'm concerned, this is the best bit about the pub/sub model - you can just swap bits in and out as needed.

Finally... I'll try to be a prompt as I can responding to issues raised on GitHub, even if they're just questions. I find it too distracting to have IRC open the whole time... but feel free to spread the word in IRC that we're here to help - people just need to let us know what their questions, bugs, features requests are, etc and we'll do our best to help them !

**Comment by draperd** [comment]
This will be available in the 1.0.36 release. You can suppress requests for user preferences for the LoggingService by configuring it so that the "suppressUserPreferences" is true (the default is false for backwards compatibility)
